app/models/concerns/labware_creators/donor_pooling_validator.rb

```diff
@@ -0,0 +1,121 @@
+  # duplicated, an error is added to the :source_barcodes attribute.
+  #
+  # @return [void]
+  def source_barcodes_must_be_different
+    return if minimal_barcodes.size == minimal_barcodes.uniq.size
+
+    errors.add(:source_barcodes, SOURCE_BARCODES_MUST_BE_DIFFERENT)
+  end
+
+  # Validates that all source plates corresponding to the minimal barcodes exist.
+  # If the number of source plates does not match the number of minimal barcodes,
+  # an error is added to the :source_plates attribute.
+  #
+  # @return [void]
+  def source_plates_must_exist
+    return if source_plates.size == minimal_barcodes.size
+
+    errors.add(:source_plates, SOURCE_PLATES_MUST_EXIST)
+  end
+
+  # Validates that the number of calculated pools does not exceed the configured
+  # number of pools. If the number of calculated pools is greater, an error is
+  # added to the :source_plates attribute.
+  #
+  # @return [void]
+  def number_of_pools_must_not_exceed_configured
+    # Don't add this error if there are already errors about missing donor_ids.
+    invalid_wells_hash = locations_with_missing_donor_id(source_plates)
+    return if invalid_wells_hash.any?
+
+    return if pools.size <= number_of_pools
+
+    errors.add(:source_plates, format(NUMBER_OF_POOLS_MUST_NOT_EXCEED_CONFIGURED, pools.size, number_of_pools))
+  end
+
+  # Validates that all wells with aliquots must have a donor_id.
+  # It uses the locations_with_missing_donor_id method to find any wells that are
+  # missing a donor_id. If any such wells are found, it adds an error message to
+  # the source_plates attribute, formatted with the barcodes of the plates and
+  # the wells that are missing a donor_id.
+  #
+  # @return [void]
+  def wells_with_aliquots_must_have_donor_id
+    invalid_wells_hash = locations_with_missing_donor_id(source_plates)
+    return if invalid_wells_hash.empty?
+
+    formatted_string = invalid_wells_hash.map { |barcode, locations| "#{barcode}: #{locations.join(', ')}" }.join(' ')
+    errors.add(:source_plates, format(WELLS_WITH_ALIQUOTS_MUST_HAVE_DONOR_ID, formatted_string))
+  end
+
+  private
+
+  # Checks each well in each source plate for missing donor_id. Returns a hash
+  # with keys as the barcodes of source plates and values as arrays of well
+  # locations with missing donor_id. If a plate has no wells with missing
+  # donor_id, it is not included in the returned hash. This method is used by
+  # the wells_with_aliquots_must_have_donor_id method to generate an error message.
+  #
+  # @param source_plates [Array] An array of source plates to check.
+  # @return [Hash] A hash mapping source plate barcodes to arrays of invalid wells.
+  def locations_with_missing_donor_id(source_plates)
+    source_plates.each_with_object({}) do |source_plate, hash|
+      invalid_wells = source_plate.wells.select { |well| missing_donor_id?(well) }
+      hash[source_plate.human_barcode] = invalid_wells.map(&:location) if invalid_wells.any?
+    end
+  end
```

Comment:
Why pass in source_plates as an argument, when you could just access it directly in the method?

Reply:
Removed the redundant param.